Physics baseline

St4rman opened the pull request:
Added in Physics from the end of the game tech module. Collisions for capsules, AABB with every col volume and OBB with every coll vol is added. Can code review Capsules but the way it currently works should just work with any capsules you use as a collision. 

Will explain if required but its just collision code.  
 :]

St4rman committed:
Updated Collision Detection Class to have all collisions

Added capsule collisions, OBB colls from previous sem physcis engine

St4rman committed:
Elasticity and systems

phases and elasticity

JunkyX1122 committed:
Imported collision lines from my (Erick) project

JunkyX1122 committed:
fun fact: nothing debug related works

JunkyX1122 committed:
Woops wrong pull request

JunkyX1122 committed:
Merge pull request #6 from The-Newcastle-Gang/DebugDrawBounding

Debug draw bounding

DelCherub commented on CSC8503CoreClasses/PhysicsSystem.cpp:
```diff
@@ -202,6 +234,70 @@ so that objects separate back out.
+	Transform& transformA = a.GetTransform();
+	Transform& transformB = b.GetTransform();
+
+	float cRestitution = physA->GetElasticity() / physB->GetElasticity();
```
Dividing coefficients of restitution isn't commutative, so the order of collision processing might make physics unreliable. Probably best to change this to something that is commutative: multiplying, averaging etc.

St4rman replied:
Noted, can get an average done. This was done for the coursework because it needed that, will replace it by the average of the two.
Cheers.

St4rman committed:
Averaging out the restitution

as per matthew's suggestion

JunkyX1122 committed:
F3 to toggle collision volume display

IMPORTANT: Currently there is a bug that prevents the amount of lines being drawn (evident when too many spheres)

JunkyX1122 committed:
isDebugDrawingCollision false by default

Joerseven commented on CSC8503CoreClasses/CollisionDetection.cpp:
```diff
@@ -49,23 +49,124 @@ bool CollisionDetection::RayIntersection(const Ray& r,GameObject& object, RayCol
+
+	Vector3 tVals(-1, -1, -1);
+
+	for (int i = 0; i < 3; ++i) {                                   //this should leave us with the best 3 intersections
```
Comments shouldn't be on same line

St4rman replied:
done

Joerseven commented on CSC8503CoreClasses/CollisionDetection.cpp:
```diff
@@ -49,23 +49,124 @@ bool CollisionDetection::RayIntersection(const Ray& r,GameObject& object, RayCol
-	return false;
+
+	Vector3 capCenter = worldTransform.GetPosition();
+	Quaternion capRota = worldTransform.GetOrientation().Normalised();
```
Rot or Rotation lmao

St4rman replied:
rota is fine

Joerseven commented on CSC8503CoreClasses/CollisionDetection.cpp:
```diff
@@ -85,10 +186,12 @@ bool CollisionDetection::ObjectIntersection(GameObject* a, GameObject* b, Collis
 	VolumeType pairType = (VolumeType)((int)volA->type | (int)volB->type);
 
 	//Two AABBs
+	//done
```
?

St4rman replied:
done

Joerseven commented on CSC8503CoreClasses/CollisionDetection.cpp:
```diff
@@ -155,41 +283,386 @@ bool CollisionDetection::AABBTest(const Vector3& posA, const Vector3& posB, cons
+	return collided;
 }
 
+bool CollisionDetection::OBBAABBIntersection(const NCL::OBBVolume& volumeA, const NCL::CSC8503::Transform& worldTransformA,
```
redundant namespaces

St4rman replied:
fixed

Joerseven commented on CSC8503CoreClasses/CollisionDetection.cpp:
```diff
@@ -155,41 +283,386 @@ bool CollisionDetection::AABBTest(const Vector3& posA, const Vector3& posB, cons
+	return collided;
+}
+
+Vector3* CollisionDetection::CalculateEdgeNormal(const Transform& inTransform,
```
Use std::array or std::vector of Vectors

St4rman replied:
OBB OBB collisions are something we are not worried about since multiple OBBS colliding in a scene should be avoided

Joerseven commented on CSC8503CoreClasses/CollisionDetection.cpp:
```diff
@@ -155,41 +283,386 @@ bool CollisionDetection::AABBTest(const Vector3& posA, const Vector3& posB, cons
+Vector3* CollisionDetection::CalculateEdgeNormal(const Transform& inTransform,
+	const Transform& inBTransform) {
+
+	Vector3* edgeNormals = new Vector3[15];
```
Corrosponding delete? Memory leak

Joerseven replied:
there is a delete lower, I am doing this with Ed and I asked him if there was a delete and he said no so not my fault.

St4rman replied:
blind

Joerseven commented on CSC8503CoreClasses/CollisionDetection.cpp:
```diff
@@ -155,41 +283,386 @@ bool CollisionDetection::AABBTest(const Vector3& posA, const Vector3& posB, cons
+
+}
+
+Vector3* CollisionDetection::CalculateVerticesOBB(const OBBVolume& inVolume, const Transform& inTransform) {
```
Same here, use std::array and get rid of vector3* and the new

St4rman replied:
OBB vs OBB should not be colliding together often since its taxing and doesnt need to happen for our game

Joerseven commented on CSC8503CoreClasses/CollisionDetection.cpp:
```diff
@@ -155,41 +283,386 @@ bool CollisionDetection::AABBTest(const Vector3& posA, const Vector3& posB, cons
 	const OBBVolume& volumeB, const Transform& worldTransformB, CollisionInfo& collisionInfo) {
-	return false;
+
+	Vector3* VolumeEdgeNormals = CalculateEdgeNormal(worldTransformA, worldTransformB); //Get axis of A and B
```
These will be replaced with std::array

Joerseven commented on CSC8503CoreClasses/CollisionDetection.h:
```diff
@@ -70,50 +70,61 @@ namespace NCL {
 		static Vector3		UnprojectScreenPosition(Vector3 position, float aspect, float fov, const Camera &c);
 		static Matrix4		GenerateInverseProjection(float aspect, float fov, float nearPlane, float farPlane);
 		static Matrix4		GenerateInverseView(const Camera &c);
+		static Vector3*		CalculateEdgeNormal(const Transform& inTransform, const Transform& inBTransform);
```
change name to edge normals

St4rman replied:
done

Joerseven commented on NCLCoreClasses/Maths.cpp:
```diff
@@ -38,12 +38,115 @@ namespace NCL {
 	
 
-		Vector3 Clamp(const Vector3& a, const Vector3&mins, const Vector3& maxs) {
+		Vector3 Clamp(const Vector3& a, const Vector3 &mins, const Vector3& maxs) {
```
Replace the dead clamp function with Ed's clamps functions

St4rman replied:
done

Joerseven reviewed (changes requested):
Comments made, generally good though 👍

St4rman committed:
PR review pass one

St4rman committed:
Merge branch 'master' into physics-baseline

St4rman commented:
Changes made. Cheers

Joerseven committed:
Fixed dirty sleep

Joerseven reviewed (approved):
LGTM!
